### PR TITLE
Configurable retention period for invalid API keys

### PR DESCRIPTION
**Pull request by ywangd**
Invalid API keys, either expired or actively invalidated, are removed from the security index eventually. However expired and invalidated keys have different retention periods. It is hard-coded to be 7 days for expired keys and Zero for invalidated keys.

This PR makes the retention period consistent for both expired and invalidated keys. It also expose a new setting so that the retention period is configurable (defaults to 7 days).

Relates: #91738

### CODE REVIEW

**Comment by ywangd** [comment]
@elasticmachine update branch

**Review by n1v0lg — APPROVED** [review]
The implementation LGTM!

One thing I wanted to discuss further: this change may introduce performance degradation for end-users that have many API keys and invalidate them frequently. Before this change,  invalidated keys are continuously deleted from the security index for such end-users, preventing the index from growing (too much). After this change, and especially if they are unaware of the new default behavior, invalidated API keys will accumulate for an entire week. 

I don't think this is a blocker since users can solve this by changing the setting. We could make this more "backwards compatible" (I'm using air quotes because I don't think this is a breaking change, per se) by adding a second setting along the lines of `delete.no_retention_on_invalidation` and default it to `true`. However, this complicates things for most users, and for us. Just wondering what your thoughts were on this trade-off. 

Approving since if we decide to stick with this behavior, the PR is good to go.

File: docs/changelog/92219.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+pr: 92219
+summary: Configurable retention period for invalid API keys
```
[the remark this answers is not in the document]
Comment (ywangd) [reply docs/changelog/92219.yaml:2]: I expanded it to say "invalidated or expired" which is still reasonably concise and argubaly more precise.

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/ApiKeyIntegTests.java [hunk 2]
```diff
@@ -459,12 +469,23 @@ private void verifyInvalidateResponse(
...
+    }
+
+    private void doTestInvalidKeysImmediatelyDeletedByRemover() throws Exception {
+        assertThat(deleteRetentionPeriodDays, equalTo(0L));
         Client client = waitForExpiredApiKeysRemoverTriggerReadyAndGetClient().filterWithHeader(
             Collections.singletonMap("Authorization", basicAuthHeaderValue(ES_TEST_ROOT_USER, TEST_PASSWORD_SECURE_STRING))
         );
 
-        List<CreateApiKeyResponse> createdApiKeys = createApiKeys(2, null).v1();
+        createApiKeys(1, TimeValue.timeValueMillis(1)).v1().get(0);
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/ApiKeyIntegTests.java:486]: Yep, don't even need `v1()` either. Removed both.

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/ApiKeyIntegTests.java [hunk 3]
```diff
@@ -568,30 +597,79 @@ public void testExpiredApiKeysBehaviorWhenKeysExpired1WeekBeforeAnd1DayBefore()
...
+                "api_key_invalidated",
+                true
+            )
+            .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+            .get();
+        assertThat(updateResponse.getResult(), is(DocWriteResponse.Result.UPDATED));
+
+        // 7th key will be deleted because it has old style invalidation (no invalidation time)
+        updateResponse = client.prepareUpdate(SECURITY_MAIN_ALIAS, createdApiKeys.get(6).getId())
+            .setDoc("api_key_invalidated", true)
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/ApiKeyIntegTests.java:663]: Added

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ExpiredApiKeysRemover.java [hunk 4]
```diff
@@ -37,17 +36,17 @@
  * Responsible for cleaning the invalidated and expired API keys from the security index.
  */
 public final class ExpiredApiKeysRemover extends AbstractRunnable {
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ExpiredApiKeysRemover.java]: I didn't rename it exactly because what you have noticed. But I don't have strong opinion about it. So I renamed it as you suggested.

**Comment by ywangd** [comment]
> this change may introduce performance degradation for end-users that have many API keys and invalidate them frequently. Before this change, invalidated keys are continuously deleted from the security index for such end-users, preventing the index from growing (too much). After this change, and especially if they are unaware of the new default behavior, invalidated API keys will accumulate for an entire week.

I consider the current behaviour of "no retention period for invalidated API keys" as a bug. There have been a few support cases where users wanted to know more about some invalidated API keys. Those keys were used by some remote clients (likely fleet agents) and caused lots of server side warning messages. But since the API keys are completely removed from the system, there was nothing we can really help. With the 7 day retention period, there is a good chance that we can pull the invalidated key's document and give out more information about it (including potentially identify the client based on the API key's metadata). No retention period is also a general security concern because it means a user can create an API key, do something with it and invalidate it immediately afterwards to make the actions essentially untraceable. Audit logging can in theory help in this case, but it is _not_ always enabled (e.g. ESS) and is a licensed feature. Therefore I think the "behaviour change" is justified. 

As for performance degradation, there can be two aspects:
1. Increased security index size - I don't think it would cause any noticeable performance difference. It is unlikely that retaining an extra week's API key documents would cause issues on the read side. It would have been an ingestion (creation) side issue first. The issue with ESS's security cluster was accumulated over _years_.
2. Increased response size of GetApiKey API - This has a better chance to be a real issue if the cluster (1) invalidate API keys very often and (2) receives many GetApiKey API calls similar to ESS's security cluster. The workaround is to use the QueryAPiKey API or configure a lower retention period. But I suspect this is *not* a normal usage pattern outside of the security cluster. 

> I don't think this is a blocker since users can solve this by changing the setting. We could make this more "backwards compatible" (I'm using air quotes because I don't think this is a breaking change, per se) by adding a second setting along the lines of delete.no_retention_on_invalidation and default it to true. However, this complicates things for most users, and for us. Just wondering what your thoughts were on this trade-off.

I don't personally want to add another setting to retain the existing behaviour by default. Because this would defeat the intention of fixing it as a bug and continue to be clueless for support cases similar to the above.

In summary, I am for keeping this PR as is. But I'd also like to know opinions from @jakelandis and @tvernum before proceeding further. Thanks!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ExpiredApiKeysRemover.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ExpiredApiKeysRemover.java]: Cheers!

**Comment by tvernum** [comment]
I'll give some high level thoughts. When it comes to the finer details, I think that's something that should be decided within the ES-Security team.

1. Every feature could break something for someone (https://xkcd.com/1172/). It's good to consider under what scenarios that might happen, but the mere existence of a hypothetical scenario isn't enough to say that a change is "breaking". We need to be able to improve the product even if the change is "only" an improvement for 99.99% of people, and has a mild negative impact on 0.01%. As an example, many performance improvements have some edge case where they're actually slower but we definitely choose to make such changes in minors. Fixing bugs often has a small negative impact on performance (just adding validation checks has a CPU impact). That's OK. 
2. One thing we should ask is "if someone does run into this hypothetical scenario, what options do they have". It sounds like they can just set the retention period to `0d`, so if there is a cluster (or a few clusters) out there that have this issue, the fix is fairly easy. It would be better for them not to have to do anything, but if we think the scenario is so rare that it probably won't happen, then it's probably OK to say "but if it does, it's just a matter of changing 1 setting".  
3. There's a delicate tradeoff between getting the right behaviour for new clusters, and not breaking existing clusters. We don't have a good story there because any new setting/change in default value affects all clusters equally, whether they're brand new, or upgraded. In theory we could decide that we want the retention period to be `0d` for existing clusters, and `7d` for new clusters but that's hard to do with the way our settings work.
4. Settings that exist just to disable other settings are terrible. In this case, if we wanted to default to no retention (which I'm not encouraging) it would be to just default the retention period to `0d` rather than add another setting.
5. The existing deletion logic is problematic. We've known that for some time, and we need some path to fixing that.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java [hunk 5]
```diff
@@ -163,6 +163,11 @@ public class ApiKeyService {
         TimeValue.timeValueHours(24L),
         Property.NodeScope
     );
+    public static final Setting<TimeValue> DELETE_RETENTION_PERIOD = Setting.positiveTimeSetting(
+        "xpack.security.authc.api_key.delete.retention_period",
+        TimeValue.timeValueDays(7),
+        Property.NodeScope
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:169]: Should this by dynamic? My gut says yes.

Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java]: Good point! Making it dynamic definitely helps with the concern of "breaking behaviour". I pushed https://github.com/elastic/elasticsearch/pull/92219/commits/bcd84d9560d13ae5457fcfcae7a03726bc7e7ca0 for it.

Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java]: It could be a better user story if we also make `delete.interval` dynamic. It should be a separate PR though.

**Comment by tvernum** [comment]
I'm not planning on re-reviewing (the previous review was really just a drive-by based on the ping about "breaking").
Let me know if you feel you need my review.

**Comment by ywangd** [comment]
@tvernum It's all good. Your view about the "breaking" discussion is all I was after. No need for full review. There is no fun in reading the tests :) Thanks! I'll be merging this soon.
